### ISSUE TITLE
Support for promises

### ISSUE THREAD

**Issue by MatiasOlivera** [mikeerickson/validatorjs#303]
## tl;dr
Use promises instead of the actual API

## Motivations
- Use promises instead of a "proprietary" API
- Decouple and separate the implementation of the public API
- Use the same API for sync and async validations
- Use control flow primitives (`then/catch` or `try/catch`)
- Better error handling
- Be able to use `async `functions

## Proposed changes

**Use promises instead of a "proprietary" API**

- It is a native and well-known API
- There is no need to learn another API

```js
const Validator = require('validatorjs');

const rules = { firstName: 'required' };
const data = { firstName: '' };

const validation = new Validator(rules, data);

// success
validation.passes();

// failure
if (validation.fails()) {
  const errors = validation.errors.all();
}

```
```js
const Validator = require('validatorjs');

const rules = { firstName: 'required' };
const data = { firstName: '' };

const validator = new Validator(rules);

validator
  .validate(data)
  .then(() => {
    // success
  })
  .catch(errors => {
    // failure
  });

```

---

**Use the same API for sync and async validations**

- There is no need to use/maintain two different APIs
- And there is no need to worry about using one or another according to the validation rules

```js
// Asynchronous api
if (validator.hasAsync) {
  validator.passes(() => {
    // success
  });
  validator.fails(() => {
    // failure
    const errors = validator.errors.all();
  });
}

// Synchronous api
if (validator.passes()) {
  // success
} else {
  // failure
  const errors = validator.errors.all();
}

```
```js
// Async and sync api
validator
  .validate(data)
  .then(() => {
    // success
  })
  .catch(errors => {
    // failure
  });

```

---

**Be able to use `async `functions**
```js
const validateData = async data => {
  try {
    await validator.validate(data);
  } catch (errors) {
    return errors;
  }
};

```

## Workaround

Of course, you can wrap the actual API with a promise, but it is not an official API

```js
const Validator = require("validatorjs");

class MyValidator {
  validate(rules, data) {
    return new Promise((resolve, reject) => {
      const validator = new Validator(data, rules);
      const handleFails = () => reject(validator.errors.all());

      // Asynchronous handler (with callbacks)
      if (validator.hasAsync) {
        validator.passes(() => resolve());
        validator.fails(() => handleFails());
      } else {
        // Synchronous handler
        if (validator.passes()) {
          resolve();
        } else {
          handleFails();
        }
      }
    });
  }
}

const rules = { firstName: "required" };
const data = { firstName: "John" };

const validator = MyValidator.validate(rules, data);
validator.catch(errors => errors);
```

## Related issues
- #106 Rewrite and refactoring
- #192 Support of async/await
- #253 Support for async/await

## Questions

- What do you think?
- What things would you improve?
- Are there plans to implement this feature?

**Comment by sanonz** [comment]
I'm not sure this is a bug?
```js
if (validator.hasAsync) {
  // This will be checked twice
  validator.passes(() => resolve());
  validator.fails(() => handleFails());

  // resolve
  validator.checkAsync(() => resolve(), () => handleFails());
} else {
  // ...
}
```
Both methods call the check method `checkAsync()` (`passes()` and `fails()`):
https://github.com/skaterdav85/validatorjs/blob/master/src/validator.js#L492
https://github.com/skaterdav85/validatorjs/blob/master/src/validator.js#L506

In the case of checking if a user exists, the query database executes the SQL twice.

**Comment by MatiasOlivera** [comment]
@sanonz It's a bug. Look at the issue #168

**Comment by sanonz** [comment]
@MatiasOlivera ok.
